### PR TITLE
Add clang_format_tool to tools/

### PR DESCRIPTION
**Pull request by noc0lour**
Imported the tools used by MongoDB developers to clean up their tree with some tweaks already for the GR source tree.

Using #1662 as base it will format all (.h|cc|c) files according to the `.clang-format` defined it the root directory of the source tree.

I'm not yet happy with the commandline interface and some intermediate functions have to be improved to pass my arbitrary "I can live with
that type of Python" standard.

I want to show in which direction repo tooling can go and get you guys involved/informed early.

### CODE REVIEW

**Comment by noc0lour** [comment]
Need to test this tool with Python2 now, since I'm developing on a Python3-centric machine

**Comment by noc0lour** [comment]
Also I'd like to have a `tools/` directory with tools not only useful for binary packaging GNU Radio but all sorts of tools. (e.g. maintainance, check-patch, ...)

File: tools/clang_format.py [hunk 1]
```diff
@@ -42,27 +50,15 @@
 #
 
 # Expected version of clang-format
-CLANG_FORMAT_VERSION = "3.8.0"
-CLANG_FORMAT_SHORT_VERSION = "3.8"
+CLANG_FORMAT_VERSION = "5.0.1"
+CLANG_FORMAT_SHORT_VERSION = "5.0"
```
Comment (marcusmueller) [inline tools/clang_format.py]: So, this is OK because you know we're running this only on machines with cool distros, right?

**Comment by marcusmueller** [comment]
Are you sure https://github.com/gnuradio/gnuradio/pull/1663/commits/5969e380c99d1934ef793fdd2defdfef679947c3 should be part of this PR?

File: tools/clang_format.py (hunk 1 above)
Comment (noc0lour) [reply tools/clang_format.py]: I need to figure out which version of clang-format we want to dictate, so formatting will be the same for all contributors.

**Comment by noc0lour** [comment]
The commit is necessary if you want to run the tool and compile the result afterwards.

File: tools/clang_format.py (hunk 1 above)
Comment (marcusmueller) [reply tools/clang_format.py]: well, that's then going to be a problem – different packagers package different versions. I wonder if we can have something like a "reformatting" push remote, which lets you push, then wait a couple second, then lets you pull, giving you a `--fixup` commit with the reformatted / `clang-tidy`ied files.

Comment (noc0lour) [reply tools/clang_format.py]: Across all major distributions you have access to at least two or more compiler major versions.

Comment (noc0lour) [reply tools/clang_format.py]: I's 4.0.1 now, which I believe we can expect people to have available.

**Comment by noc0lour** [comment]
Using `python3` it should be sane now to run `python3 ./tools/clang_format.py format` using this branch and look at the diff. clang-format needs to run twice due to grave issues in qtgui code. 
Then `python3 ./tools/clang_format lint` should remain silent.

If python3 is your default python no prefixing of python3 is needed for executing the tool.

**Comment by noc0lour** [comment]
Branch reformatting (updating a branch created before tree reformatting happened) is not yet working.

**Comment by noc0lour** [comment]
This tool has two major modes:
 `format` - will format files in place
 `lint` - will show what is not up to the standard defined in `.clang-format`

`format` also has a special  "rebase" mode `-b` which will update a branch based on the commit before the codebase reformatting to base on the commit which did the formatting. `-b $hash-before $hash-after`.

File: tools/clang_format.py (hunk 1 above)
Comment (mbr0wn) [reply tools/clang_format.py]: Agreed. FYI, we did the same for UHD.

**Comment by mbr0wn** [comment]
Once #1664 is merged, this can be rebased and merged too, right? I'm fine with the state. It's all new code, it can't break anything.

**Pull request merged**